### PR TITLE
Extract country information from phone number

### PR DESCRIPTION
**Pull request by renatofilho**
During the contact import if the country field is empty try to extract
that information from phone number

### CODE REVIEW

File: tests/auto/client/test_contactsimporter.cpp [hunk 1]
```diff
@@ -151,8 +151,8 @@ private Q_SLOTS:
         const QVector<ContactsSet> contacts = importer.contacts();
         QCOMPARE(contacts.size(), 1);
         const SugarAccount account = contacts.at(0).account;
-        // auto fill country based on phone number with the system locale
-        QCOMPARE(account.billingAddressCountry(), QString::fromUtf8("Estados Unidos"));
+        // Make sure that the country name still on english even if the app Locale is not
```
Comment (dfaure-kdab) [inline tests/auto/client/test_contactsimporter.cpp]: still on english -> is still in English

File: client/src/utilities/contactsimporter.cpp [hunk 2]
```diff
@@ -162,3 +182,24 @@ QVector<ContactsSet> ContactsImporter::contacts() const
...
+
+    const QStringList localeName = QLocale().name().split(QStringLiteral("_"));
+    auto phoneUtil = i18n::phonenumbers::PhoneNumberUtil::GetInstance();
+    i18n::phonenumbers::PhoneNumber number;
+    if (phoneUtil->Parse(phone.toStdString(), localeName.last().toStdString(), &number) == i18n::phonenumbers::PhoneNumberUtil::NO_PARSING_ERROR) {
+        // use ZZ region to make sure that always will return country name
+        const std::string description = i18n::phonenumbers::PhoneNumberOfflineGeocoder().GetDescriptionForNumber(number,
+                                                                                                                 icu::Locale("en", "US"),
+                                                                                                                 i18n::phonenumbers::RegionCode::ZZ());
+        return QString::fromStdString(description).split(QStringLiteral(",")).last().trimmed();
```
Comment (dfaure-kdab) [inline client/src/utilities/contactsimporter.cpp]: I find split overkill for just a lastIndexOf + mid, no point in creating and filling a temporary container. But yes I'm nitpicking, performance doesn't matter in this code, so feel free to ignore that ;)

File: client/src/utilities/contactsimporter.cpp [hunk 3]
```diff
@@ -162,3 +182,24 @@ QVector<ContactsSet> ContactsImporter::contacts() const
...
+    } else {
+        qWarning() << "Fail to parse numnber" << phone;
```
Comment (dfaure-kdab) [inline client/src/utilities/contactsimporter.cpp]: typo: number

File: tests/auto/client/test_contactsimporter.cpp [hunk 4]
```diff
@@ -125,6 +125,50 @@ private Q_SLOTS:
...
+        QVERIFY(importer.importFile(file.fileName()));
+        const QVector<ContactsSet> contacts = importer.contacts();
+        QCOMPARE(contacts.size(), 1);
+        const SugarAccount account = contacts.at(0).account;
+        QCOMPARE(account.name(), QString("KDAB"));
+        // auto fill country based on phone number
+        QCOMPARE(account.billingAddressCountry(), QString::fromUtf8("France"));
+    }
+
+    void testImportingAccountWithDiffLocale()
```
Comment (dfaure-kdab) [inline tests/auto/client/test_contactsimporter.cpp]: a data-driven unittest would be really appropriate here, I think.

File: client/src/utilities/contactsimporter.cpp [hunk 5]
```diff
@@ -162,3 +182,24 @@ QVector<ContactsSet> ContactsImporter::contacts() const
...
     return mContacts;
 }
+
+QString ContactsImporter::extractCountryFromPhoneNumber(const QString &phone) const
+{
+    if (phone.isEmpty()) {
+        return QString();
+    }
+
+    const QStringList localeName = QLocale().name().split(QStringLiteral("_"));
```
[the remark this answers is not in the document]
Comment (renatofilho) [reply client/src/utilities/contactsimporter.cpp]: fixed

File: client/src/utilities/contactsimporter.cpp [hunk 6]
```diff
@@ -162,3 +182,24 @@ QVector<ContactsSet> ContactsImporter::contacts() const
...
+    auto phoneUtil = i18n::phonenumbers::PhoneNumberUtil::GetInstance();
+    i18n::phonenumbers::PhoneNumber number;
+    if (phoneUtil->Parse(phone.toStdString(), localeName.last().toStdString(), &number) == i18n::phonenumbers::PhoneNumberUtil::NO_PARSING_ERROR) {
+        // use ZZ region to make sure that always will return country name
```
[the remark this answers is not in the document]
Comment (renatofilho) [reply client/src/utilities/contactsimporter.cpp]: fixed

File: client/src/utilities/contactsimporter.cpp (hunk 2 above)
Comment (renatofilho) [reply client/src/utilities/contactsimporter.cpp]: fixed

File: client/src/utilities/contactsimporter.cpp (hunk 3 above)
Comment (renatofilho) [reply client/src/utilities/contactsimporter.cpp]: fixed

File: tests/auto/client/test_contactsimporter.cpp (hunk 1 above)
Comment (renatofilho) [reply tests/auto/client/test_contactsimporter.cpp]: fixed

File: tests/auto/client/test_contactsimporter.cpp (hunk 4 above)
Comment (renatofilho) [reply tests/auto/client/test_contactsimporter.cpp]: fixed

File: client/src/CMakeLists.txt [hunk 7]
```diff
@@ -130,9 +130,16 @@ target_link_libraries(fatcrmprivate
...
+    set(OptionalLibraries
+        ${PhoneNumber_LIBRARIES}
+        ${PhoneNumberGeoCoding_LIBRARIES}
+        ${Libicu_LIBRARIES}
+     )
+endif()
 target_link_libraries(fatcrmprivate
   ${KDReports_LIBRARIES}
+  ${OptionalLibraries}
 )
```
Comment (krf) [inline client/src/CMakeLists.txt]: Easier:
```
if (PhoneNumber_FOUND)
    target_link_libraries(fatcrmprivate
        ${PhoneNumber_LIBRARIES}
        ...
```

File: cmake/FindLibicu.cmake [hunk 8]
```diff
@@ -0,0 +1,26 @@
+# - Find Libicu
```
Comment (krf) [inline cmake/FindLibicu.cmake]: CMake provides a find script: FindICU.cmake -- does that help?

File: client/src/utilities/contactsimporter.cpp [hunk 9]
```diff
@@ -31,6 +31,42 @@
 #include <QFile>
 #include "fatcrm_client_debug.h"
 #include <QTextCodec>
+#include <QLocale>
+#include <QStringList>
+#include <QString>
+
+#include "config-phonenumber.h"
+
+#if USE_PHONENUMBER
```
Comment (dfaure-kdab) [inline client/src/utilities/contactsimporter.cpp]: Indenting inside #if is very strange and unusual ;)

IIRC an indented #include statement isn't even portable to all compilers.

File: cmake/FindPhoneNumber.cmake [hunk 10]
```diff
@@ -0,0 +1,31 @@
+# - Find PhoneNumber
+# This module finds if PhoneNumber is installed.
+#
+# PhoneNumber_FOUND                 - Set to TRUE if PhoneNumber was found.
+# PhoneNumber_LIBRARIES             - Path to PhoneNumber libraries.
+# PhoneNumber_INCLUDE_DIR           - Path to the PhoneNumber include directory.
+#
```
Comment (dfaure-kdab) [inline cmake/FindPhoneNumber.cmake]: .. and PhoneNumberGeoCoding_LIBRARIES

**Pull request merged**